### DISCUSSION TITLE
Dockerfile option Volume - where is it mounted with k3s?

### DISCUSSION THREAD

**Discussion by sando38** [k3s-io/k3s#5770] [Q&A] [answered]
Hi friends,

I have a quick question. In a Dockerfile I can add something like this:

```yaml
VOLUME ["$HOME/log", "$HOME/run"]
```

If I start the container with `docker run ...` docker will create volumes in the path `/var/lib/docker/volumes` with a random name. However, when I use the same image to start a container in k3s, I am unable to find any volume on the host machine. 

So, are those "volumes" mounted somewhere else?

Hope you could follow my thoughts. If anything is written unclear, please let me know. 

Have a great evening,

sando

**Comment by brandond** [comment]
This is more of a Kubernetes question, see the discussion here: https://github.com/kubernetes/kubernetes/issues/52032

**Comment by ChristianCiach** [comment, marked as answer]
Yes, the volumes are still mounted, but (as you expected) somewhere else.

K3s bundles containerd as the container runtime instead of docker. You can use the command `k3s ctr` as a CLI for containerd. 

As an example, let's take a look at the official [redis image](https://hub.docker.com/_/redis) which declares a volume in its Dockerfile:

```
VOLUME [/data]
```

Let's use `k3s ctr container list` (which is roughly equivalent to `docker ps`) to find the container-id. Keep in mind that you have to run this command on the node where the container is actually running. For me it looks like this:

```
e029be12bf574fb3c8ae226a50e5428a0f68cebfe8c18da9af9a840d7706bf99    docker.io/library/redis@sha256:21db12e5ab3cc343e9376d655e8eabbdbe5516801373e95a8a9e66010c5b8819              io.containerd.runc.v2
```

Alright. Now we take the container-id and feed it to `k3s ctr container info`, which is roughly equivalent to `docker inspect`. The output is a json document that contains this element:
```json
        "mounts": [
            {
                "destination": "/data",
                "type": "bind",
                "source": "/var/lib/rancher/k3s/agent/containerd/io.containerd.grpc.v1.cri/containers/e029be12bf574fb3c8ae226a50e5428a0f68cebfe8c18da9af9a840d7706bf99/volumes/cabd5615b60c408b24700549002ad0b9b609c529481e955a9143919c8d7bc56a",
                "options": [
                    "rbind",
                    "rprivate",
                    "rw"
                ]
            },
```

There you go. It looks like containerd creates the volumes at `{{ k3s-data-dir }}/agent/containerd/io.containerd.grpc.v1.cri/containers/{{ container-id }}/volumes/{{ volume-id }}`.

**Reply by brandond** [reply]
Note that this is currently an containerd implementation detail; neither kubernetes nor the CRI spec require that they go anywhere in particular. The upstream issue I linked to tracks nailing some of this down a bit more to ensure that the 'implicit volumes' created via the Dockerfile `VOLUME` instruction can be tracked more easily.

**Comment by sando38** [comment]
Thanks a lot to you both! This morning I also thought, well, actually a containerd related question.. Nevertheless, thanks!
